predicate/predicate.py

```diff
@@ -60,33 +24,62 @@ def __init__(self, expr):
         self.field = None
 
     def get_field(self, instance):
-        lookup_type = 'exact' # Default lookup type
-        parts = self.lookup.split(LOOKUP_SEP)
-        num_parts = len(parts)
-        if (len(parts) > 1 and parts[-1] in QUERY_TERMS):
-            # Traverse the lookup query to distinguish related fields from
-            # lookup types.
-            lookup_model = instance
-            for counter, field_name in enumerate(parts):
+        """
+        Returns the value of this lookup on the given instance.
+
+        If the path ends with a value having the same name as a lookup
+        type (like "year", "range", etc.), the lookup must provide an
+        explicit type, such as "__exact".
+
+        Lookups can span Django fields, dicts, lists, or other integer-
+        indexed iterables.
+        """
+
+        def traverse_attr(o, attr, reraise=False):
+            # Convert numbers to integers to allow list indexing
+            if attr.isdigit():
+                attr = int(attr)
+
+            # M2M relations are a special case
+            if isinstance(o, Manager):
+                if isinstance(attr, int):
+                    return o.all()[attr]
```

Comment:
First off, thanks for the thoughtful review and for helping out with the project! I'm trying to put myself back in my mental space from 2 years ago as I read through the code and comments.

One concrete example I can think of is a related manager configured with something like `order_by = ('-priority',)`. Then I might want a predicate where the first item is or is not type `BUG`.

Mainly though I think it exists for the principle of least surprise. If it works on a list, then there's no reason I as a consumer of the API would expect that it doesn't work on a relationship that is modeled as a list.

I agree it is brittle.